### PR TITLE
 added oleds 

### PR DESCRIPTION
**Pull request by daysgobye**
Hello, another pr on the docket for you lol. This is my oled extension, I have been spending a lot of time on this and I don't think I need to make any more changes in the near future so it's time to let others play with it. 

### peg prefix
you will see that it's prefixed with Peg. Peg is some software I am working on to let you configure your kmk board. The prefix is to denote that this is the OLED extension that peg interacts with. Anyone can use it and make changes to it but if there were 2 OLED extensions I wanted people to know that they are different. This does not seem to be a problem the last OLED extension never got pulled in. Being that not only users will interact with this but also other code some of my design decisions were about that. I have however tried to make this easy to use for the end-user.

### Docs
I kinda did something a little different in the docs by including images. But I wanted to try and make the best docs I could.

### Code
So one of the biggest things I focused on with this extension is to make it work on nrf52xx chips most chips people seem to be using have more rom&ram so if it can fit on the n!n it can work on anything. 

On that note, you will see my GC calls often I have tested the extension quite a bit to make sure it is stable. The next thing I wanted to make sure I did was to make this expandable, I want to add other features as time goes on. I feel like I have a solid base here that we can expand on.  You will see in "returnCurrectRenderText" my plan. I want to be able to do custom logic for each type of reaction and with this layout, I should be able to. WPM might need to do the math here or read a pin for battery percentage. But the user code stays the same in field 0 tell it what to react to > in field 1 gives it the data to display (images or text)

#### how we do images
Due to the memory constraints, I was not able to get the images working in the more traditional manner and have to load them right from the disk. This works well and saves us some ram.

### CODE REVIEW

**Comment by kdb424** [comment]
Unlike your other module, this is much less invasive, and OLED modules will likely never be so generic to work with absolutely everything. Once tests pass, I'll take a final look at code, but considering your use case, I'm fine having a semi-specialized OLED module like this in the repo.

File: docs/peg_oled_display.md
Comment (klardotsh) [inline docs/peg_oled_display.md:62]: API change: this should not need to take `keyboard` as an init arg, because the Modules and Extensions APIs already pass the keyboard object reference as the first argument to `during_bootup` (https://github.com/KMKfw/kmk_firmware/blob/a210c2bf62c15c70e6af4c42d5b8bf6984bece5b/kmk/kmk_keyboard.py#L422-L424 and https://github.com/KMKfw/kmk_firmware/blob/a210c2bf62c15c70e6af4c42d5b8bf6984bece5b/kmk/kmk_keyboard.py#L429-L431). Initialization routines that require access to the overall keyboard object should be able to defer that access until `on_bootup` is called.

File: docs/peg_oled_display.md
Comment (klardotsh) [inline docs/peg_oled_display.md:80]: weird markdown nit: while this reads correctly in raw source, when this MD file is rendered to any other format, it will be interpreted as a numbered list and screw up the indexes (into 1-4). suggest formatting as

```
- element 0: top left
- element 1: top right
...
```

File: docs/peg_oled_display.md
Comment (klardotsh) [inline docs/peg_oled_display.md:56]: users should not have to pass magic strings to enable functionality, especially capitalized strings. this uses excess RAM and is extremely error-prone/brittle for end users. enumerated values should be integer enum-like objects/singletons imported from the extension as `from peg_oled_display import EntityToDisplay` (or some such wording), for example https://github.com/KMKfw/kmk_firmware/blob/a210c2bf62c15c70e6af4c42d5b8bf6984bece5b/kmk/extensions/rgb.py#L71-L80

File: docs/peg_oled_display.md
Comment (klardotsh) [inline docs/peg_oled_display.md:53]: users should not have to pass magic strings to enable functionality, especially capitalized strings. this uses excess RAM and is extremely error-prone/brittle for end users. enumerated values should be integer enum-like objects/singletons imported from the extension as `from peg_oled_display import ReactionType`, for example https://github.com/KMKfw/kmk_firmware/blob/a210c2bf62c15c70e6af4c42d5b8bf6984bece5b/kmk/extensions/rgb.py#L71-L80

File: docs/peg_oled_display.md
Comment (klardotsh) [inline docs/peg_oled_display.md:50]: I'm making these comments in reverse order, I'm now realizing, but fine, whatever: if the data is this structured, it probably shouldn't be an anonymous array. it should be a class of its own outright. (in desktop Python I'd also suggest a `namedtuple` but we no longer use those here because their implementation [sometimes takes an external dependency](https://github.com/KMKfw/kmk_firmware/pull/203#discussion_r654896112))

File: kmk/extensions/peg_oledDisplay.py
Comment (klardotsh) [inline kmk/extensions/peg_oledDisplay.py:20]: these inline type hints have to go in comment form rather than inline, sadly. the PEP that implements inline type hints isn't universally implemented in CircuitPython (last we checked in https://github.com/KMKfw/kmk_firmware/pull/226#issuecomment-903316710)

File: kmk/extensions/peg_oledDisplay.py
Comment (klardotsh) [inline kmk/extensions/peg_oledDisplay.py:27]: since these are never used again (and tuples aren't negligible in memory), don't store these on `self` in tuple form. storing the int representation is fine, but the tuples should be passed to the SSD1306 constructor and then allowed to be garbage collected (or whatever SSD1306 does to them)

**Review by klardotsh — DISMISSED** [review]
Other notes:

- rename `peg_oledDisplay.py` -> `peg_oled_display.py`
- bonus points if we can get a desktop unit test for any of this, but it's not a blocker
- @xs5871 and @kdb424 can dismiss my review at basically any point to unblock merge; I'm going to be out of town all of next week and not on GitHub and don't want to hold this up

**Comment by daysgobye** [comment]
@klardotsh Thank you for looking it over I agree on 9/10 points I'm going to fix it up .

**Comment by daysgobye** [comment]
I think it got it all changed over very good changes I am happy with them.

File: docs/peg_oled_display.md
Comment (kdb424) [inline docs/peg_oled_display.md:7]: I'm being picky, but can we capitalize the lists? It just looks better, and is more consistent with the rest of the docs I believe. This applies to all start of lines that start a new sentence, or bullet point.

**Review by kdb424 — CHANGES_REQUESTED** [review]
Once the changes to the docs formatting is done, it's good by me. I'll let XS have a week or two to nitpick as they are busy, but otherwise, I won't block a merge pending those fixes.

File: docs/peg_oled_display.md
Comment (daysgobye) [reply docs/peg_oled_display.md]: good catch!

File: docs/peg_oled_display.md
Comment (kdb424) [inline docs/peg_oled_display.md:17]: Missed a few

File: docs/peg_oled_display.md
Comment (kdb424) [inline docs/peg_oled_display.md:26]: This as well. Any time there is a new line, follow English rules :p

File: docs/peg_oled_display.md
Comment (kdb424) [inline docs/peg_oled_display.md:36]: This one

File: docs/peg_oled_display.md
Comment (kdb424) [inline docs/peg_oled_display.md:54]: I'll let this one go because it's a technical word

File: docs/peg_oled_display.md
Comment (kdb424) [inline docs/peg_oled_display.md:56]: You get the idea

File: docs/peg_oled_display.md
Comment (daysgobye) [reply docs/peg_oled_display.md]: let me know if you dont like my solution

**Review by xs5871 — COMMENTED** [review]
I don't see a reason to block. It's documented and self-contained; even if it is special purpose, it may either differentiate in the future, or provide a reference for others.
And @klardotsh did a better job at nitpicking than I would have.

**Comment by daysgobye** [comment]
Thank you xs. Are we missing anything?

**Comment by xs5871** [comment]
Merge is still blocked. I think @klardotsh and @kdb424 need to re-review or approve the changes?

**Review by xs5871 — APPROVED** [review]

**Comment by xs5871** [comment]
I'm sorry, I don't know how I can mark the change requests as resolved, other than dismissing them.

**Comment by kdb424** [comment]
Seems reasonable to me. Thanks for the review and merge!

**Pull request merged**
